usr/src/uts/common/fs/zfs/zfs_dir.c

```diff
@@ -609,16 +609,23 @@ zfs_rmnode(znode_t *zp)
 	}
 
 	/*
-	 * Free up all the data in the file.
+	 * Free up all the data in the file.  We don't do this for directories
+	 * because we need truncate and remove to be in the same tx, like in
+	 * zfs_znode_delete(). Otherwise, if we crash here we'll end up with
+	 * an inconsistent truncated zap object in the delete queue.  Note a
```

Comment:
@behlendorf Suggested doing on normal directory. The reason is that we don't want to leave with an invalid ZAP object on disk.